analyses.go

```diff
@@ -627,6 +627,58 @@ func getTimeLimit(ctx context.Context, dedb *sql.DB, analysisID string) (int64,
 	return timeLimitSeconds, nil
 }
 
+// EnsureSubdomain makes sure the provided job has a subdomain set in the DB, returning it
+func EnsureSubdomain(ctx context.Context, dedb *sql.DB, analysis *Job) (string, error) {
+	if analysis.Subdomain == "" {
+		// make sure to use analysis.ID, not external ID here.
+		userID, err := getUserIDForJob(ctx, dedb, analysis.ID)
+		if err != nil {
+			return "", errors.Wrapf(err, "error getting userID for job %s", analysis.ID)
+		} else {
+			log.Infof("user id is %s and invocation id is %s", userID, analysis.ExternalID)
+
+			// make sure to use externalID, not analysis.ID here
+			subdomain := generateSubdomain(userID, analysis.ExternalID)
+
+			log.Infof("generated subdomain for analysis %s is %s, based on user ID %s and invocation ID %s", analysis.ID, subdomain, userID, analysis.ExternalID)
+
+			if err = setSubdomain(ctx, dedb, analysis.ID, subdomain); err != nil {
+				return "", errors.Wrapf(err, "error setting subdomain for analysis '%s' to '%s'", analysis.ID, subdomain)
+			}
+			return subdomain, nil
+		}
+	} else {
+		return analysis.Subdomain, nil
+	}
+}
+
+func EnsurePlannedEndDate(ctx context.Context, dedb *sql.DB, analysis *Job) error {
+	// Check to see if the planned_end_date is set for the analysis
+	if analysis.PlannedEndDate != "" {
+		log.Infof("planned end date for %s is set to %s, nothing to do", analysis.ID, analysis.PlannedEndDate)
+		return nil // it's already set, so move along.
+	}
+
+	startDate, err := time.ParseInLocation(TimestampFromDBFormat, analysis.StartDate, time.Local)
```

Comment:
I _believe_ this is the correct thing to do here. The timestamp in the DB for the start date is a timezone-less timestamp but stored using our local time zone, not UTC, as far as I know. So this may have previously been setting wholly incorrect planned end dates.